runtime/compiler/trj9/env/VMJ9.cpp

```diff
@@ -1269,21 +1269,47 @@ TR_OpaqueClassBlock *
 TR_J9VMBase::getObjectClass(uintptrj_t objectPointer)
    {
    TR_ASSERT(haveAccess(), "Must haveAccess in getObjectClass");
-   if (TR::Compiler->om.generateCompressedObjectHeaders())
-      return (TR_OpaqueClassBlock*)((uintptrj_t)*(uint32_t*)(objectPointer + TR::Compiler->om.offsetOfObjectVftField()) & TR::Compiler->om.maskOfObjectVftField());
+   return convertClassPtrToClassOffset(J9OBJECT_CLAZZ(vmThread(), objectPointer));
+   }
+
+uintptrj_t
+TR_J9VMBase::getStaticReferenceFieldAtAddress(TR::Compilation* comp, uintptrj_t fieldAddress, TR::SymbolReference *symRef)
+   {
+   TR_ASSERT(haveAccess(), "Must haveAccess in getStaticReferenceFieldAtAddress");
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+   // Emit read barrier
+   if (TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads())
+      {
+      return (uintptrj_t)J9STATIC_OBJECT_LOAD(vmThread(), symRef->getOwningMethod(comp)->containingClass(), fieldAddress);
+      }
    else
-      return (TR_OpaqueClassBlock*)(*(uintptrj_t*)(objectPointer + TR::Compiler->om.offsetOfObjectVftField()) & TR::Compiler->om.maskOfObjectVftField());
+#endif
+      {
+      return *(uintptrj_t *)fieldAddress;
+      }
+   }
+
+uintptrj_t
+TR_J9VMBase::getReferenceAtAddress(uintptrj_t address)
+   {
+   TR_ASSERT(haveAccess(), "Must haveAccess in getReferenceFieldAtAddress");
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+   // Emit read barrier
+   if (TR::Compiler->om.shouldGenerateReadBarriersForFieldLoads())
+      vmThread()->javaVM->memoryManagerFunctions->J9ReadBarrier(vmThread(), (fj9object_t *)address);
+#endif
+
+   return *(uintptrj_t*)address;
```

Comment:
same thing here, just call J9OBJECT_OBJECT_LOAD (which will do the barrier and the de-reference)

Reply:
J9OBJECT_OBJECT_LOAD would do de-compressing of the address, so that above comment applies to getReferenceFieldAtAddress. I'm a bit confused with helper....

Reply:
It looks like `J9ReadBarrier` is used for possibly compressed references, as they occur within objects, but this function is used for uncompressed references only

Reply:
Ok, I realized that J9OBJECT_OBJECT_LOAD takes an offset (as opposed to an absolute address of the field), so it could not be used to implement getReferenceFieldAtAddress. 

But I'm still confused by this helper. So it returns potentially a compressed value (fj9object_t). It's typically not how VM and GC code works, where locals and method arguments (for heap object addresses) are in decompressed form. Only slots in heap would really be in compressed form. But if this works for you, ok...

Still, there is concern for the input argument. By the analogy, I'm guessing it would also be in compressed form, but it's not what J9ReadBarrier expects as an address - it expects uncompressed form.

Reply:
It looks like in all places where I use this helper, previously it was just a simple dereference. @jdmpapin and I had a discussion about it, and it seems like that's ok because they're JNI refs, which are always decompressed. Therefore, the use of `J9ReadBarrier` here is definitely incorrect. 

With that said however, we still need a way of dereferencing a pointer to an uncompressed object ref. After looking through the implementation of `J9STATIC_OBJECT_LOAD`, it looks like the `J9Class` parameter is never used. This brings up two questions:

1. Can we pass in NULL for the J9Class parameter?
2. Can we use J9STATIC_OBJECT_LOAD to dereference a pointer to an uncompressed object ref?

Reply:
1) you can as far as Concurrent Scavenger read barrier is concerned. It would work now, but not necessarily in future.
2) It would work, but definitely it would be abuse of API. We can create a new one (which would indeed do the same thing).

You don't have to introduce the barrier for JNI slots, yet. VM knows about all of them and they are treated as hard roots in CS (will be updated during root scanning in first STW phase), or if they are overwritten meanwhile with another value, that value is guarantied to be already up to date (like to would be a case with any other hard root slot).

However, this is not optimal behavior. Some JNI slots could be weak refs and should not be treated as hard roots (we leak this memory in CS, till next global GC.) This is just a temporary solution in CS, before we identify all JNI slots and update the reads with appropriate API - we actually have a lot of them using raw reads even in core VM.

So, anything related to JNI slots should be a separate change, and should wait for a new API to be introduced. What we need sooner is to identify accesses to class statics.

Reply:
> you can as far as Concurrent Scavenger read barrier is concerned. It would work now, but not necessarily in future.

For clarification, would the `J9class` parameter be the class of the object who's static field we're dereferencing, or is it the class of the static object?

As @jdmpapin pointed out in a comment below:
>Here receiverChild->getSymbolReference() might not be the right symbol reference - e.g. it could be an auto.
>
>If receiverChildConstraint is a VPKnownObject instead of a VPConstString, then there might not even be a static symbol reference that corresponds to the string

The symbol reference is used to get the class by getting the owning method. This might be ok if the symref refers to an auto, but in the second case, if we don't have a symref, we won't be able to get the `J9Class`. What is the potential benefit of knowing what the `J9Class` is?

Reply:
> For clarification, would the J9class parameter be the class of the object who's static field we're dereferencing, or is it the class of the static object?

It's the former.

It's hard to speculate what it could be used in future. 

If you don't have an easy way to find the class pointer, passing null should be ok.

Reply:
> What we need sooner is to identify accesses to class statics

Are string literals counted as class statics for this purpose? Or are those hard roots as well?

> would the J9class parameter be the class of the object who's static field we're dereferencing, or is it the class of the static object? [...] The symbol reference is used to get the class by getting the owning method.

I don't believe the class of the symref's owning method is necessarily either of these. It should be the class containing the access, e.g. consider these classes:

    class Foo {
        public static final Object BAR = [...];
    }

    class Baz {
        public Object quux() {
            return Foo.BAR;
        }
    }

If we're looking at `aload Foo.BAR` feeding the return, the symref corresponds to the symbolic reference `Foo.BAR` in the constant pool of `Baz`. Its owning method should be `Baz.quux`

Reply:
> Are string literals (edit: and call site table entries, further edit: and method type table entries) counted as class statics for this purpose? Or are those hard roots as well?

Short answer is, we just need those that access J9Class::ramStatics. That's the minimum required for Concurrent Scavenger to work. I'm not completely sure why others are not needed (my claim is based on looking on Scavenger code that does class specific scanning in hard roots), but I believe all others are in Tenure space already (perhaps because they rarely or never die or get modified) and not subject to movement during CS.

Long answer is, we need all, because eventually concurrent copying logic can be part of other GC policies where any object anywhere in the heap will be subject to movement